File: src/pages/leads_list_page.py
"""Leads List page object (/lead-management/list) - includes the Analytics tab
(LeadAnalytics component, same route, not a separate page - see context/ui-context.md).
Locators from LMS.pdf visible text and cofee-web route/permission source.
"""

from playwright.sync_api import Page

from src.pages.base_page import BasePage


class LeadsListPage(BasePage):
    def __init__(self, page: Page, base_url: str) -> None:
        super().__init__(page, base_url)
        # --- List tab elements ---
        self.list_tab = self.page.get_by_role("tab", name="List")
        self.analytics_tab = self.page.get_by_role("tab", name="Analytics")
        self.search_input = self.page.get_by_role("searchbox")
        self.filter_btn = self.page.get_by_role("button", name="Filter")
        self.add_lead_btn = self.page.get_by_role("button", name="Add Lead")
        self.leads_table = self.page.get_by_role("table")
        self.no_results_message = self.page.get_by_text(
            "No leads found"
        )  # [Assumption] - exact copy unconfirmed, see matrix Open Questions

        # --- Analytics tab elements ---
        self.leads_by_top_locations_heading = self.page.get_by_text(
            "Leads By Top Locations"
        )
        self.leads_by_source_heading = self.page.get_by_text("Leads By Source")
        self.lead_count_revenue_by_status_heading = self.page.get_by_text(
            "Lead Count & Revenue By Status"
        )
        self.leads_distribution_by_agents_heading = self.page.get_by_text(
            "Leads Distribution By Agents"
        )
        self.agents_table_count_toggle = self.page.get_by_role("button", name="Count")
        self.agents_table_revenue_toggle = self.page.get_by_role(
            "button", name="Revenue"
        )
        self.no_leads_data_message = self.page.get_by_text("No Leads Data Available")

    def goto_list(self) -> None:
        self.goto("lead-management/list")

    def open_analytics_tab(self) -> None:
        self.analytics_tab.click()

    def search_leads(self, term: str) -> None:
        self.search_input.fill(term)

File: src/pages/LeadsListPage.ts
import type { Locator, Page } from '@playwright/test';

/**
 * /lead-management/list — "Leads" (List tab; Analytics is a tab on the same
 * route, confirmed live 2026-08-24 — not a separate URL).
 */
export class LeadsListPage {
  readonly page: Page;

  readonly listTab: Locator;
  readonly analyticsTab: Locator;
  readonly addLeadButton: Locator;
  readonly statusCells: Locator;

  /** Confirmed live 2026-08-24. */
  static readonly EXPECTED_COLUMN_HEADERS = [
    'Customer Name',
    'Mobile Number',
    'Tags',
    'Source',
    'Assignee',
    'Days Stuck',
    'Status',
    'Actions',
  ] as const;

  constructor(page: Page) {
    this.page = page;
    this.listTab = page.getByRole('tab', { name: 'List' });
    this.analyticsTab = page.getByRole('tab', { name: 'Analytics' });
    this.addLeadButton = page.getByRole('button', { name: 'Add Lead' });
    // Confirmed live 2026-08-24: the leads table is div-based, not a real
    // <table> — no row/table role to scope by. Matches a status *value*
    // exactly, so it can't collide with the "Status" column header text.
    this.statusCells = page.getByText(
      new RegExp(`^(${['Prospect', 'Contacted', 'Market Qualified', 'Parked', 'Sales Qualified', 'Interested', 'In Discussion', 'Onboarded', 'Won', 'Lost'].join('|')})$`),
    );
  }

  async goto(): Promise<void> {
    await this.page.goto('/lead-management/list', { waitUntil: 'domcontentloaded' });
  }

  async getColumnHeaderTexts(): Promise<string[]> {
    await this.page.getByText(LeadsListPage.EXPECTED_COLUMN_HEADERS[0], { exact: true }).first().waitFor({ state: 'visible' });

    const texts: string[] = [];
    for (const header of LeadsListPage.EXPECTED_COLUMN_HEADERS) {
      const locator = this.page.getByText(header, { exact: true }).first();
      // eslint-disable-next-line no-await-in-loop
      texts.push((await locator.isVisible()) ? header : '');
    }
    return texts;
  }
}
